src/java/org/apache/cassandra/index/sai/disk/v1/InvertedIndexSearcher.java

```diff
@@ -168,6 +168,9 @@ private Cell<?> readColumn(SSTableReader sstable, PrimaryKey primaryKey)
         var slices = Slices.with(indexContext.comparator(), Slice.make(primaryKey.clustering()));
         try (var rowIterator = sstable.iterator(dk, slices, columnFilter, false, SSTableReadsListener.NOOP_LISTENER))
         {
+            // primaryKey might not belong to this sstable, thus the iterator will be empty
+            if (rowIterator.isEmpty())
```

Comment:
Nice fix. Adding a semi-related observation. If we end up seeing a lot of object allocations associated with the `Slices` which are `O(n)` for results matched in the `WHERE` clause, it might be worth finding a way to use `sstable.couldContain(dk)`, which checks the bloom filter, to reduce allocations for keys not in the sstable. If you look in the implementation of the `iterator` method, you'll see how it already does that for us, but it's after we've already created our slice object.

Reply:
Good point. I will not be surprised if there are more places for such improvements.